### PR TITLE
Add omitempty for DisableCompression and DisableKeepAlive fields in ScrapeConfig

### PR DESCRIPTION
**Pull request by sti-jans**
Added the "omitempty" option for the DisableCompression and DisableKeepAlive fields in ScrapeConfig.

We use your models in our code and would like to be able to leave these fields empty. I think, it's not going to be a problem.

### CODE REVIEW

**Comment by valyala** [comment]
Could you describe your case? Probably it would be better using third-party package for generating Prometheus-compatible yaml configs such as https://godoc.org/github.com/percona/promconfig ?

**Comment by sti-jans** [comment]
We use VictoriaMetrics as storage and generate the configuration of our Prometheus.
You are probably right, but I would like to keep the models up-to-date for the configuration. Not all parameters are available in the repository you specified.

Perhaps the query is not good, because I added tags for only two options.

**Comment by valyala** [comment]
> Perhaps the query is not good, because I added tags for only two options.

As I understood, `omitemtpy` annotation is needed for skipping marshaling default values into yaml. If this is the intended behavior, then could you add `omitempty` annotation to all the default values in order to minimize the resulting yaml file?

**Comment by sti-jans** [comment]
@valyala I added `omitempty` annotation to all the default and the optional values.

File: lib/promauth/config.go [hunk 1]
```diff
@@ -13,18 +13,18 @@ import (
...
 type TLSConfig struct {
-	CAFile             string `yaml:"ca_file"`
-	CertFile           string `yaml:"cert_file"`
-	KeyFile            string `yaml:"key_file"`
-	ServerName         string `yaml:"server_name"`
+	CAFile             string `yaml:"ca_file,omitempty"`
+	CertFile           string `yaml:"cert_file,omitempty"`
+	KeyFile            string `yaml:"key_file,omitempty"`
+	ServerName         string `yaml:"server_name,omitempty"`
 	InsecureSkipVerify bool   `yaml:"insecure_skip_verify"`
```
Comment (valyala) [inline lib/promauth/config.go:20]: I thought that `insecure_skip_verify` is also optional

File: lib/promscrape/config.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ var (
 // Config represents essential parts from Prometheus config defined at https://prometheus.io/docs/prometheus/latest/configuration/configuration/
 type Config struct {
 	Global        GlobalConfig   `yaml:"global"`
-	ScrapeConfigs []ScrapeConfig `yaml:"scrape_configs"`
+	ScrapeConfigs []ScrapeConfig `yaml:"scrape_configs,omitempty"`
```
Comment (valyala) [inline lib/promscrape/config.go:40]: I'm unsure whether `Config` is useful without `scrape_configs` list

File: lib/promscrape/config.go [hunk 3]
```diff
@@ -102,8 +102,8 @@ type FileSDConfig struct {
 //
 // See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#static_config
 type StaticConfig struct {
-	Targets []string          `yaml:"targets"`
-	Labels  map[string]string `yaml:"labels"`
+	Targets []string          `yaml:"targets,omitempty"`
```
Comment (valyala) [inline lib/promscrape/config.go:105]: I'm unsure whether `StaticConfig` is usable without `targets` list.

File: lib/promscrape/discovery/consul/consul.go [hunk 4]
```diff
@@ -10,18 +10,18 @@ import (
...
 type SDConfig struct {
-	Server       string              `yaml:"server"`
+	Server       string              `yaml:"server,omitempty"`
 	Token        *string             `yaml:"token"`
 	Datacenter   string              `yaml:"datacenter"`
-	Scheme       string              `yaml:"scheme"`
+	Scheme       string              `yaml:"scheme,omitempty"`
 	Username     string              `yaml:"username"`
 	Password     string              `yaml:"password"`
 	TLSConfig    *promauth.TLSConfig `yaml:"tls_config"`
```
Comment (valyala) [inline lib/promscrape/discovery/consul/consul.go:19]: I think `tls_config` is optional here

File: lib/promscrape/discovery/consul/consul.go [hunk 5]
```diff
@@ -10,18 +10,18 @@ import (
...
-	Services     []string            `yaml:"services"`
-	Tags         []string            `yaml:"tags"`
+	Services     []string            `yaml:"services,omitempty"`
+	Tags         []string            `yaml:"tags,omitempty"`
 	NodeMeta     map[string]string   `yaml:"node_meta"`
```
Comment (valyala) [inline lib/promscrape/discovery/consul/consul.go:22]: It looks like `node_meta` is optional here

File: lib/promscrape/discovery/dns/dns.go [hunk 6]
```diff
@@ -17,7 +17,7 @@ import (
 // See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#dns_sd_config
 type SDConfig struct {
 	Names []string `yaml:"names"`
-	Type  string   `yaml:"type"`
+	Type  string   `yaml:"type,omitempty"`
 	Port  *int     `yaml:"port"`
```
Comment (valyala) [inline lib/promscrape/discovery/dns/dns.go:21]: `port` is optional here

File: lib/promscrape/discovery/dockerswarm/dockerswarm.go [hunk 7]
```diff
@@ -14,11 +14,11 @@ type SDConfig struct {
 	// TODO: add support for proxy_url
 	TLSConfig *promauth.TLSConfig `yaml:"tls_config"`
```
Comment (valyala) [inline lib/promscrape/discovery/dockerswarm/dockerswarm.go:15]: `tls_config` is optional

File: lib/promscrape/discovery/ec2/ec2.go [hunk 8]
```diff
@@ -8,17 +8,17 @@ import (
 //
 // See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ec2_sd_config
 type SDConfig struct {
-	Region    string `yaml:"region"`
+	Region    string `yaml:"region,omitempty"`
 	Endpoint  string `yaml:"endpoint"`
```
Comment (valyala) [inline lib/promscrape/discovery/ec2/ec2.go:12]: `endpoint` is optional

File: lib/promscrape/discovery/ec2/ec2.go [hunk 9]
```diff
@@ -8,17 +8,17 @@ import (
...
-	AccessKey string `yaml:"access_key"`
-	SecretKey string `yaml:"secret_key"`
+	AccessKey string `yaml:"access_key,omitempty"`
+	SecretKey string `yaml:"secret_key,omitempty"`
 	// TODO add support for Profile, not working atm
 	Profile string `yaml:"profile"`
```
Comment (valyala) [inline lib/promscrape/discovery/ec2/ec2.go:16]: `profile` is optional

File: lib/promscrape/discovery/ec2/ec2.go [hunk 10]
```diff
@@ -8,17 +8,17 @@ import (
...
 	RoleARN string `yaml:"role_arn"`
```
Comment (valyala) [inline lib/promscrape/discovery/ec2/ec2.go:17]: `role_arn` is optional

File: lib/promscrape/discovery/gce/gce.go [hunk 11]
```diff
@@ -13,8 +13,8 @@ type SDConfig struct {
 	Filter  string   `yaml:"filter"`
```
Comment (valyala) [inline lib/promscrape/discovery/gce/gce.go:13]: `filter` is optional

File: lib/promscrape/discovery/kubernetes/kubernetes.go [hunk 12]
```diff
@@ -10,13 +10,13 @@ import (
...
-	BearerToken     string                    `yaml:"bearer_token"`
-	BearerTokenFile string                    `yaml:"bearer_token_file"`
-	TLSConfig       *promauth.TLSConfig       `yaml:"tls_config"`
-	Namespaces      Namespaces                `yaml:"namespaces"`
+	BasicAuth       *promauth.BasicAuthConfig `yaml:"basic_auth,omitempty"`
+	BearerToken     string                    `yaml:"bearer_token,omitempty"`
+	BearerTokenFile string                    `yaml:"bearer_token_file,omitempty"`
+	TLSConfig       *promauth.TLSConfig       `yaml:"tls_config,omitempty"`
+	Namespaces      Namespaces                `yaml:"namespaces,omitempty"`
 	Selectors       []Selector                `yaml:"selectors"`
```
Comment (valyala) [inline lib/promscrape/discovery/kubernetes/kubernetes.go:20]: `selectors` is optional

File: lib/promscrape/discovery/openstack/openstack.go [hunk 13]
```diff
@@ -25,10 +25,10 @@ type SDConfig struct {
 	Region                      string `yaml:"region"`
 	// RefreshInterval time.Duration `yaml:"refresh_interval"`
 	// refresh_interval is obtained from `-promscrape.openstackSDCheckInterval` command-line option.
-	Port         int                 `yaml:"port"`
-	AllTenants   bool                `yaml:"all_tenants"`
+	Port         int                 `yaml:"port,omitempty"`
+	AllTenants   bool                `yaml:"all_tenants,omitempty"`
 	TLSConfig    *promauth.TLSConfig `yaml:"tls_config"`
```
Comment (valyala) [inline lib/promscrape/discovery/openstack/openstack.go:30]: `tls_config` is optional

File: lib/promscrape/discovery/openstack/openstack.go [hunk 14]
```diff
@@ -25,10 +25,10 @@ type SDConfig struct {
 	Region                      string `yaml:"region"`
```
Comment (valyala) [inline lib/promscrape/discovery/openstack/openstack.go:25]: All the fields except of `role` and `region` are optional here

**Review by valyala — CHANGES_REQUESTED** [review]

File: lib/promauth/config.go (hunk 1 above)
Comment (sti-jans) [reply lib/promauth/config.go]: I thought the `insecure_skip_verify` field might indicate why there are no others.
But you're probably right.

File: lib/promscrape/config.go (hunk 2 above)
Comment (sti-jans) [reply lib/promscrape/config.go]: you're right

File: lib/promscrape/config.go (hunk 3 above)
Comment (sti-jans) [reply lib/promscrape/config.go]: you're right
I fixed

File: lib/promscrape/discovery/consul/consul.go (hunk 4 above)
Comment (sti-jans) [reply lib/promscrape/discovery/consul/consul.go]: fixed

File: lib/promscrape/discovery/consul/consul.go (hunk 5 above)
Comment (sti-jans) [reply lib/promscrape/discovery/consul/consul.go]: fixed

File: lib/promscrape/discovery/ec2/ec2.go (hunk 10 above)
Comment (sti-jans) [reply lib/promscrape/discovery/ec2/ec2.go]: fixed

File: lib/promscrape/discovery/gce/gce.go (hunk 11 above)
Comment (sti-jans) [reply lib/promscrape/discovery/gce/gce.go]: fixed

File: lib/promscrape/discovery/kubernetes/kubernetes.go (hunk 12 above)
Comment (sti-jans) [reply lib/promscrape/discovery/kubernetes/kubernetes.go]: fixed

File: lib/promscrape/discovery/openstack/openstack.go (hunk 13 above)
Comment (sti-jans) [reply lib/promscrape/discovery/openstack/openstack.go]: fixed

File: lib/promscrape/discovery/openstack/openstack.go (hunk 14 above)
Comment (sti-jans) [reply lib/promscrape/discovery/openstack/openstack.go:25]: all fixed

File: lib/promscrape/discovery/dns/dns.go (hunk 6 above)
Comment (sti-jans) [reply lib/promscrape/discovery/dns/dns.go]: fixed

File: lib/promscrape/discovery/dockerswarm/dockerswarm.go (hunk 7 above)
Comment (sti-jans) [reply lib/promscrape/discovery/dockerswarm/dockerswarm.go]: fixed

File: lib/promscrape/discovery/ec2/ec2.go (hunk 8 above)
Comment (sti-jans) [reply lib/promscrape/discovery/ec2/ec2.go]: fixed

File: lib/promscrape/discovery/ec2/ec2.go (hunk 9 above)
Comment (sti-jans) [reply lib/promscrape/discovery/ec2/ec2.go]: fixed

**Review by valyala — APPROVED** [review]
LGTM

**Comment by valyala** [comment]
@sti-jans , thanks for the contribution!

**Pull request merged**